src/hyperloglog.c

```diff
@@ -0,0 +1,439 @@
+        case 6:
+            k1 ^= (uint64_t)(tail[5]) << 40;
+        case 5:
+            k1 ^= (uint64_t)(tail[4]) << 32;
+        case 4:
+            k1 ^= (uint64_t)(tail[3]) << 24;
+        case 3:
+            k1 ^= (uint64_t)(tail[2]) << 16;
+        case 2:
+            k1 ^= (uint64_t)(tail[1]) << 8;
+        case 1:
+            k1 ^= (uint64_t)(tail[0]) << 0;
+            k1 *= c1;
+            k1 = (k1 << 31) | (k1 >> (64 - 31));
+            k1 *= c2;
+            h1 ^= k1;
+    };
+
+    //----------
+    // finalization
+
+    h1 ^= nbytes;
+    h2 ^= nbytes;
+
+    h1 += h2;
+    h2 += h1;
+
+    h1 ^= h1 >> 33;
+    h1 *= 0xff51afd7ed558ccdULL;
+    h1 ^= h1 >> 33;
+    h1 *= 0xc4ceb9fe1a85ec53ULL;
+    h1 ^= h1 >> 33;
+
+    h2 ^= h2 >> 33;
+    h2 *= 0xff51afd7ed558ccdULL;
+    h2 ^= h2 >> 33;
+    h2 *= 0xc4ceb9fe1a85ec53ULL;
+    h2 ^= h2 >> 33;
+
+    h1 += h2;
+    h2 += h1;
+
+//    ((uint64_t *) retbuf)[0] = h1;
+//    ((uint64_t *) retbuf)[1] = h2;
+	return h1;
+}
+
+
+/* hash value space cardinality 2^64-1 */
+#define HASH_CARDINALITY ( (uint64_t)  (((int64_t)   ((uint64_t)  (1l << 63))) >> 63)   )
+//#define HASH_CARDINALITY ( 1l << 32)
+
+
+
+hyperloglog_t *hll_init(int p) {
+    hyperloglog_t *this = malloc(sizeof(hyperloglog_t));
+    this->m = 1 << p;
+    this->M = malloc(this->m);
+    memset(this->M, 0, this->m);
+    this->p = p;
+    return this;
+}
+
+
+double hll_error(int p) {
+    return 1.04 / sqrt(1 << p);
+}
+
+
+void hll_add(hyperloglog_t *this, void *v, size_t nbytes) {
+    uint64_t h = qhashmurmur3_64(v, nbytes);
+    uint64_t idx = h & (this->m-1);
+    uint64_t w = __builtin_clz(h | (this->m-1)) + 1;
+    this->M[idx] = this->M[idx] > w ? this->M[idx] : w;
+}
+
+double estimate_bias(double E, uint8_t p){
+	int idx = (p - MIN_P);
+    double * red =RAW_ESTIMATE_DATA[idx];
+    double * bd = BIAS_DATA[idx];
+    int red_length = (int) red[0];
+    if(E <= red[1]) {
+      return bd[1];
+    }
+    if(E > red[red_length - 1]) {
+      return 0.0;
+    }
+    for(int i = 1; i < red_length - 2; i++) {
+      if(red[i] < E && E < red[i + 1]) {
+        double slope = (bd[i + 1] - bd[i])/(red[i + 1] - red[i]);
+        double intercept = bd[i + 1] - (slope * red[i + 1]);
+        return (slope * E) + intercept;
+      }
+    }
+    return 0.0;
+}
+
+double hll_estimate(hyperloglog_t *this) {
+    double sum = 0;
+    int zeros = 0;
+
+    /* Calculate the third term of the HLL equation:
+     /      p               \ `      
+     |  __ 2   -  1   - M[i]| - 1 
+     | \            2       |     
+     \ /__ i = 0            /     */
+     
+    //Calculate amount of zeros in M as well for the last step
+    for (int i = 0; i < this->m; i++) {
+        sum += 1.0 / (1 << this->M[i]);
+
+        if (this->M[i] == 0)
+            zeros++;
+    }
+
+    // second term approximations from the google paper 
+    // http://tech.adroll.com/blog/data/2013/07/10/hll-minhash.html
```

Comment:
Nitpick but comments style in this file is not consistent with itself and the rest of the code.